### ISSUE TITLE
Allow syntactic rule application inside a lexical rule

### ISSUE THREAD

**Issue by pdubroy** [harc/ohm#136]
As discussed on [this thread](https://groups.google.com/a/cdglabs.org/forum/#!topic/ohm/Vl3PBFn6g58) in the Google Group:

> Take JS's template strings, for example:
> 
> ```
> `The answer is ${6 * 7}`
> ```
> 
> To parse this, it would be nice if you could write something like:
> 
> ```
> tstr
>   = "`" tstrPart* "`"
> 
> tstrPart
>   = "${" Expr "}"      -- expr
>   | ~("`" | "${") any  -- char
> ```
> 
> Right now you can't do this, because that invocation of Expr is not allowed. But it could be.
> 
> Yesterday Pat and I talked about this, and we think that it might be a good idea to allow syntactic rules to be invoked from inside lexical rules -- this would probably require an operator, something like the analog of the "lexification" operator, `#`. For now, I'll use `@`.
> 
> We could rewrite the tstrPart rule above using @ as follows:
> 
> ```
> tstrPart
>   = "${" @Expr "}"      -- expr
>   | ~("`" | "${") any   -- char
> ```
> 
> The semantics of `@` would be the same as the semantics of the application of a start rule that is syntactic. In other words, `@Expr` would skip both leading and trailing whitespace. (That's exactly what you'd want in this example.)

**Comment by codeZeilen** [comment]
Why would you need a special operator for this? Is there any other semantic that an application of a syntactic rule might have at that point?

(Just stumbled upon this as we noticed an inconsistency in Ohm/S which does not restrict users to use lexical and syntactic rules in the bodies of rules of the other kind.)

**Comment by pdubroy** [comment]
@codeZeilen Just seeing this comment now, since this issue came up on the Ohm Discord server. I can't actually remember why we wanted to introduce special syntax for this.

@alexwarth, do you remember? Do you see any problem with allowing syntactic rules to be appear in the body of a lexical rule, without any explicit syntax?

**Comment by 4silvertooth** [comment]
> issue came up on the Ohm Discord server.

Here is the grammar

```
TapeGrammar {

Tape 
  = ( #lineComment | Assign | ResultBlock | #op )+
  | end
  
ResultBlock
  = TapeBlock #(result)?

TapeBlock
  = TapeBlock #addop #atom   -- add
  | TapeBlock  #subop  #atom   -- subtract	
  | TapeBlock  #mulop  #atom   -- multiply
  | TapeBlock  #divop   #atom   -- divide
  | TapeBlock  #addop  #atomPerc   -- addPerc
  | TapeBlock  #subop  #atomPerc   -- subtractPerc	
  | TapeBlock  #mulop  #atomPerc   -- multiplyPerc
  | TapeBlock  #divop   #atomPerc   -- dividePerc
  | atom
  
atom 
  = (number | sid) ~perc #comment?
  
atomPerc 
  = (number | id) perc #comment?

Assign
  = id eqop (number | id) (&eol | comment)?
 
comment
  = spacesNoNL* ~eol letter (~eol ~op any)* (&eol | &op | end)

lineComment
  = (eol+ | letter) letter (~eol ~"=" any)* (&eol | end)

number (a number)
  = sign? spacesNoNL* digit* "." digit+   -- decimal
  | sign? spacesNoNL* digit+              -- integer

addop   
  = spacesNoNL* eol? "+"
  
subop
  = spacesNoNL* eol? "-"
    
mulop
  = spacesNoNL* eol? ("*"|"×")

divop   
  = spacesNoNL* eol? ("/"|"÷")
  
perc
  = "%"

eqop
  = "="

sign
  = ("+" | "-")

spacesNoNL 
  = ~eol "\x00".." "

op
  = (addop | subop | mulop | divop)

result
  = spacesNoNL* eol (eqop spacesNoNL* number)?

sid
  = sign id

id 
  = letter alnum*

signId
  = sign id

eol 
	= "\r\n" | "\n" | "\r" | "\u2028" | "\u2029"  

}
```
for input, 
```
         100
+  200
+     300
```
with the `entry rule` `Tape` being a `Syntactic rule` the spaces before `100` are skipped, 


![test1](https://user-images.githubusercontent.com/793967/119438080-089b9200-bd3d-11eb-8800-26abf68843bc.png)


while spaces before the other numbers `200` and `300` are counted for, 


![test2](https://user-images.githubusercontent.com/793967/119438094-1224fa00-bd3d-11eb-9351-d204c1b670ed.png)


my semantics when ever encounters rule `op` needs to format the input to justify the numbers with proper spaces
```
    100
+   200
+   300
```

and if the `entry rule` itself is made to be a `lexical rule`, we _can't use any syntactic rule through out the grammar_.

The result is 


![result](https://user-images.githubusercontent.com/793967/119439860-da1fb600-bd40-11eb-9938-a208496a6b69.gif)


while all other number rule gets indented, not the first line.
This is an attempt to make an opensource alternative to https://calctape.app/

**Comment by alexwarth** [comment]
Sorry for the late reply, it's been a crazy week.

> @alexwarth, do you remember? Do you see any problem with allowing syntactic rules to be appear in the body of a lexical rule, without any explicit syntax?

@pdubroy yes, I do. Consider the following grammar:

```
a = b C d
C = ...
```

If we allow lexical rule `a` to invoke syntactic rule `C`, we have to decide what happens with the implicit skipping of whitespace. We may decide to skip spaces:

(A) before and after `C`, or
(B) only before `C`, or
(C) not at all.

I have a strong preference for (A), but that's not the point that I'm trying to make here.

No matter what option we go with, we should try to avoid surprising the programmer. And I feel that allowing lexical rules to invoke syntactic rules without any special syntax is asking for trouble, because it hides the fact that something tricky is going on. (One could argue that the fact that `C` is written in uppercase is enough of a visual hint, but they would be wrong :).)

If it were only up to me, I would repurpose the "lexification operator" `#` to serve as a portal between the lexical and syntactic worlds. That change would be backwards-compatible, i.e., it would preserve the meaning of `#` when used inside a syntactic rule. And when used inside a lexical rule, it would enable people to do the sort of thing that was discussed at the top of this thread:

```
tstr
  = "`" tstrPart* "`"

tstrPart
  = "${" #Expr "}"     -- expr
  | ~("`" | "${") any  -- char
```

I'd also be happy (maybe equally happy) with the `@` syntax that we came up with when we first thought of this issue a few years ago. But simply allowing syntactic rules to be invoked inside lexical rules is asking for trouble...

**Comment by 4silvertooth** [comment]
I am no expert at this but, both  Expr and #Expr inside lexical rules should be valid syntactic rule,
If one doesn't want to skip the spaces before the `Expr` 
`
tstr
  = "`" tstrPart* "`"

tstrPart
  = "${" #Expr "}"     -- expr
  | ~("`" | "${") any  -- char
`

and

`
tstr
  = "`" tstrPart* "`"

tstrPart
  = "${" Expr "}"     -- expr
  | ~("`" | "${") any  -- char
`
if the the spaces before `Expr` are to be skipped.

**Comment by pdubroy** [comment]
@alexwarth Thanks for the details. I remembered that it was about space-skipping, and my thinking was that the uppercase first letter is enough of a hint. But, now that I think about it more I can see that having a special syntax would help ensure that people don't do this accidentally, but instead consciously use it only when it's required.

About what the syntax should be, I'm not sure about reusing `#` for this, for a couple reasons:

- the meaning is context-dependent, which could be confusing when developing or refactoring a grammar.
- it seems more complex and abstract to have a "portal operator" rather than a "lexification operator", and more difficult to explain

Another idea that I think we discussed at one point: what about using `!` as a kind of "yes, I know what I'm doing" operator? This might be useful for other cases where we want to disallow things by default, but still have an escape hatch.

This is a bit similar to Ruby's [bang methods](https://thoughtfulapps.com/articles/ruby/about-ruby-bang-methods) (which I assume were inspired by Scheme).

So, the example you gave would look like this:

```
tstrPart
  = "${" Expr! "}"     -- expr
  | ~("`" | "${") any  -- char
```

**Comment by pdubroy** [comment]
@4silvertooth That's an interesting case, but I don't think we need to support it. You can always use `~space` inside a lexical context to disallow spaces (and prevent space skipping) at that position. If we had the bang operator that I described above:

```
tstrPart
  = "${" ~space Expr! "}"  -- expr
  | ~("`" | "${") any      -- char
```

...should do the trick.

**Comment by pdubroy** [comment]
This has shipped in an experimental form in v16.1.0-pre.1, which can be installed via `npm install ohm-js@next`. It's also available in the Ohm Editor by appending `?ohm-js=next` to the URL.

Here is a sample grammar, using the `tstr` example that has been used in this thread:

```
G {
  tstr = "`" tstrPart* "`"

  tstrPart
    = "${" experimentalApplySyntactic<Expr> "}"  -- expr
    | ~("`" | "${") any  -- char
  
  Expr = number "+" number
  number = digit+
}
```

Note that both leading and trailing whitespace will be skipped around _Expr_. Here is a screenshot from the editor with _Show spaces_ enabled:

<img width="317" alt="Screen Shot 2021-11-03 at 12 48 01" src="https://user-images.githubusercontent.com/17293/140055246-b1634924-94a6-4a6d-a057-7e96863c2b2e.png">

Feedback on this is very much welcome!

**Comment by guitarvydas** [comment]
feedback: working fine for me (https://github.com/guitarvydas/jam0002/blob/main/guitarvydas/cmm.ohm)

**Comment by pdubroy** [comment]
Thanks for the feedback, @guitarvydas!

FYI I've renamed the rule to `applySyntactic` and released it as part of [Ohm v16.1.0](https://github.com/harc/ohm/discussions/346).
